module/applications/journal/class-summary-sheet.mjs

```diff
@@ -0,0 +1,439 @@
+        { content: game.i18n.localize("JOURNALENTRYPAGE.DND5E.Class.SpellSlotLevel") }
+      ]];
+      table.cols = [{class: "spellcasting", span: 2}];
+
+      // Loop through each level, gathering "Spell Slots" & "Slot Level" for each one
+      for ( const level of Array.fromRange(CONFIG.DND5E.maxLevel, 1) ) {
+        actor.items.get(classId).updateSource({"system.levels": level});
+        actor.reset();
+        table.rows.push([
+          { class: "spell-slots", content: `${actor.system.spells.pact.max}` },
+          { class: "slot-level", content: actor.system.spells.pact.level.ordinalString() }
+        ]);
+      }
+    }
+
+    // Leveled Progression
+    else {
+      // Get the max spell slot size based on progression
+      actor.items.get(classId).updateSource({"system.levels": CONFIG.DND5E.maxLevel});
+      actor.reset();
+      const largestSlot = Object.entries(actor.system.spells).reduce((slot, [key, data]) => {
+        if ( data.max === 0 ) return slot;
+        const level = parseInt(key.replace("spell", ""));
+        if ( !Number.isNaN(level) && level > slot ) return level;
+        return slot;
+      }, -1);
+
+      // Prepare headers & columns
+      table.headers = [
+        [{content: game.i18n.localize("JOURNALENTRYPAGE.DND5E.Class.SpellSlotsPerSpellLevel"), colSpan: largestSlot}],
+        Array.fromRange(largestSlot, 1).map(spellLevel => ({content: spellLevel.ordinalString()}))
+      ];
+      table.cols = [{class: "spellcasting", span: largestSlot}];
+
+      // Loop through each level, gathering max slots for each level
+      for ( const level of Array.fromRange(CONFIG.DND5E.maxLevel, 1) ) {
+        actor.items.get(classId).updateSource({"system.levels": level});
+        actor.reset();
+        const cells = [];
+        for ( const spellLevel of Array.fromRange(largestSlot, 1) ) {
+          const max = actor.system.spells[`spell${spellLevel}`]?.max;
+          cells.push({class: "spell-slots", content: max || "&mdash;"});
+        }
+        table.rows.push(cells);
+      }
+    }
+
+    return table;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare options table based on optional GrantItem advancement.
+   * @param {Item5e} item    Class item belonging to this journal.
+   * @returns {object|null}  Prepared optional features table.
+   */
+  async _getOptionalTable(item) {
+    const headers = [[
+      { content: game.i18n.localize("DND5E.Level") },
+      { content: game.i18n.localize("DND5E.Features") }
+    ]];
+
+    const cols = [
+      { class: "level", span: 1 },
+      { class: "features", span: 1 }
+    ];
+
+    const makeLink = uuid => TextEditor._createContentLink(["", "UUID", uuid]).outerHTML;
+
+    const rows = [];
+    for ( const level of Array.fromRange(CONFIG.DND5E.maxLevel, 1) ) {
+      const features = [];
+      for ( const advancement of item.advancement.byLevel[level] ) {
```

Comment:
We could consider factoring this out to share code with `_getTable`, similar to what we've done with `_getFeatures`, or is that maybe premature if we plan to add additional feature types here?

Reply:
This will eventually also handle ability score improvement & subclass advancements in the main table but not the optional table, but there might still be a way to factor out some of this code.